core/chains/evm/gas/rollups/l1_oracle.go

```diff
@@ -0,0 +1,313 @@
+	gasPriceMethod      string
+	l1GasPriceMethodAbi abi.ABI
+	l1GasPriceMu        sync.RWMutex
+	l1GasPrice          *assets.Wei
+
+	l1GasCostAddress   string
+	gasCostMethod      string
+	l1GasCostMethodAbi abi.ABI
+
+	chInitialised chan struct{}
+	chStop        services.StopChan
+	chDone        chan struct{}
+}
+
+const (
+	// ArbGasInfoAddress is the address of the "Precompiled contract that exists in every Arbitrum chain."
+	// https://github.com/OffchainLabs/nitro/blob/f7645453cfc77bf3e3644ea1ac031eff629df325/contracts/src/precompiles/ArbGasInfo.sol
+	ArbGasInfoAddress = "0x000000000000000000000000000000000000006C"
+	// ArbGasInfo_getL1BaseFeeEstimate is the a hex encoded call to:
+	// `function getL1BaseFeeEstimate() external view returns (uint256);`
+	ArbGasInfo_getL1BaseFeeEstimate = "getL1BaseFeeEstimate"
+	// NodeInterfaceAddress is the address of the precompiled contract that is only available through RPC
+	// https://github.com/OffchainLabs/nitro/blob/e815395d2e91fb17f4634cad72198f6de79c6e61/nodeInterface/NodeInterface.go#L37
+	ArbNodeInterfaceAddress = "0x00000000000000000000000000000000000000C8"
+	// ArbGasInfo_getPricesInArbGas is the a hex encoded call to:
+	// `function gasEstimateL1Component(address to, bool contractCreation, bytes calldata data) external payable returns (uint64 gasEstimateForL1, uint256 baseFee, uint256 l1BaseFeeEstimate);`
+	ArbNodeInterface_gasEstimateL1Component = "gasEstimateL1Component"
+
+	// OPGasOracleAddress is the address of the precompiled contract that exists on OP stack chain.
+	// This is the case for Optimism and Base.
+	OPGasOracleAddress = "0x420000000000000000000000000000000000000F"
+	// OPGasOracle_l1BaseFee is a hex encoded call to:
+	// `function l1BaseFee() external view returns (uint256);`
+	OPGasOracle_l1BaseFee = "l1BaseFee"
+	// OPGasOracle_getL1Fee is a hex encoded call to:
+	// `function getL1Fee(bytes) external view returns (uint256);`
+	OPGasOracle_getL1Fee = "getL1Fee"
+
+	// ScrollGasOracleAddress is the address of the precompiled contract that exists on Scroll chain.
+	ScrollGasOracleAddress = "0x5300000000000000000000000000000000000002"
+	// ScrollGasOracle_l1BaseFee is a hex encoded call to:
+	// `function l1BaseFee() external view returns (uint256);`
+	ScrollGasOracle_l1BaseFee = "l1BaseFee"
+	// ScrollGasOracle_getL1Fee is a hex encoded call to:
+	// `function getL1Fee(bytes) external view returns (uint256);`
+	ScrollGasOracle_getL1Fee = "getL1Fee"
+
+	// GasOracleAddress is the address of the precompiled contract that exists on Kroma chain.
+	// This is the case for Kroma.
+	KromaGasOracleAddress = "0x4200000000000000000000000000000000000005"
+	// GasOracle_l1BaseFee is the a hex encoded call to:
+	// `function l1BaseFee() external view returns (uint256);`
+	KromaGasOracle_l1BaseFee = "l1BaseFee"
+
+	// Interval at which to poll for L1BaseFee. A good starting point is the L1 block time.
+	PollPeriod = 12 * time.Second
+
+	// RPC call timeout
+	queryTimeout = 10 * time.Second
```

Comment:
Not sure, but is it possible to use a configured RPC timeout like other components are doing?

Reply:
Ideally let's reuse something that already exists

Reply:
I tried to look through existing configs and didn't find one for RPC timeouts. Even in `chains/evm/client`, we have 10s [hard coded](https://github.com/smartcontractkit/chainlink/blob/139fb1af7323d573aec6a3f23f7cb49a4d13a678/core/chains/evm/client/client.go#L28) which is used by `ContextWithDefaultTimeout`. It makes sense for this to be configurable but I think I'd have to add a new config for this.

Reply:
Actually I see [RequestTimeout](https://github.com/smartcontractkit/chainlink/blob/139fb1af7323d573aec6a3f23f7cb49a4d13a678/core/config/docs/chains-starknet.toml#L10-L11) for Starknet which I could maybe leverage. But it may be misleading if the rest of our code isn't using this config for RPC timeouts.

Reply:
Yeah, [QueryTimeout](https://github.com/smartcontractkit/chainlink/blob/develop/common/client/node.go#L21) or keep was the config I was thinking. You can probably use `evmclient.ContextWithDefaultTimeout()` here instead of having a new config and let the client handle the timeout.

Reply:
I updated the code to use the `client.QueryTimeout`. I was avoiding using `ContextWithDefaultTimeout` since it just uses `context.Background()` under the hood and we have a ctx passed into the method we could use.